ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateMutators.java

```diff
@@ -103,6 +107,83 @@ public void initiateValidatorExit(MutableBeaconState state, int index) {
                     exitQueueEpoch.plus(specConfig.getMinValidatorWithdrawabilityDelay())));
   }
 
+  public void initiateValidatorExit(
+      final int index, final Supplier<ValidatorExitContext> validatorExitContextSupplier) {
+
+    final ValidatorExitContext validatorExitContext = validatorExitContextSupplier.get();
+
+    final Validator validator = validatorExitContext.validators.get(index);
+    // Return if validator already initiated exit
+    if (!validator.getExitEpoch().equals(FAR_FUTURE_EPOCH)) {
+      return;
+    }
+
+    if (validatorExitContext.exitQueueChurn.compareTo(validatorExitContext.churnLimit) >= 0) {
+      validatorExitContext.exitQueueEpoch = validatorExitContext.exitQueueEpoch.increment();
+      validatorExitContext.exitQueueChurn = UInt64.ONE;
+    } else {
+      validatorExitContext.exitQueueChurn = validatorExitContext.exitQueueChurn.increment();
+    }
+
+    // Set validator exit epoch and withdrawable epoch
+    validatorExitContext.validators.set(
+        index,
+        validator
+            .withExitEpoch(validatorExitContext.exitQueueEpoch)
+            .withWithdrawableEpoch(
+                validatorExitContext.exitQueueEpoch.plus(
+                    specConfig.getMinValidatorWithdrawabilityDelay())));
+  }
+
+  public Supplier<ValidatorExitContext> createValidatorExitContextSupplier(
+      final MutableBeaconState state) {
+    return Suppliers.memoize(
+        () -> {
+          final ValidatorExitContext validatorExitContext =
+              new ValidatorExitContext(
+                  state.getValidators(), beaconStateAccessors.getValidatorChurnLimit(state));
+
+          validatorExitContext.exitQueueEpoch = UInt64.ZERO;
+
+          final List<Validator> exitedValidators = new ArrayList<>();
+
+          for (Validator validator : state.getValidators()) {
+            if (validator.getExitEpoch().equals(FAR_FUTURE_EPOCH)) {
+              continue;
+            }
+            validatorExitContext.exitQueueEpoch =
+                validatorExitContext.exitQueueEpoch.max(validator.getExitEpoch());
+            exitedValidators.add(validator);
+          }
+          validatorExitContext.exitQueueEpoch =
+              validatorExitContext.exitQueueEpoch.max(
+                  miscHelpers.computeActivationExitEpoch(
+                      beaconStateAccessors.getCurrentEpoch(state)));
+
+          validatorExitContext.exitQueueChurn =
+              UInt64.valueOf(
+                  exitedValidators.stream()
+                      .filter(v -> v.getExitEpoch().equals(validatorExitContext.exitQueueEpoch))
+                      .count());
+
+          return validatorExitContext;
+        });
+  }
+
+  public static class ValidatorExitContext {
```

Comment:
no public getters so guessing this doesn't need to be public...

Reply:
the class must be known to `AbstractBlockProcessor`